### DISCUSSION TITLE
Fool-safe handling of non-abstract rules without assignation

### DISCUSSION THREAD

**Discussion by Aluriak** [textX/textX#350] [Q&A]
Hi !

I (the fool in this situation) lost a lot of time figuring why my grammar was failing hard:

    import textx

    METAMODEL = textx.metamodel_from_str("""
    Model: expression=Expr ;

    ParenExpr: '(' Expr ')';
    Expr: ParenExpr | BaseExpr;
    BaseExpr: left=Condition operations*=Operation;
    Operation:  op=BoolOperator remaining=Condition;

    Condition: ParenExpr | RawCondition;
    RawCondition: id=Identifier op=MathOperator val=INT;

    Identifier: id=/[a-zA-Z0-9_-]+/;
    MathOperator:  op=/=|>|</;
    BoolOperator:  op=/AND|OR/;
    """)

    raw_model = METAMODEL.model_from_str('(b=3 OR c=4)')
    print(raw_model.expression)

    raw_model = METAMODEL.model_from_str('a=2 AND (b=3 OR c=4)')
    print(raw_model.expression)

On the input `a=2 AND (b=3 OR c=4)`, it would provides me correctly the python objects for `a=2`, but the string `(b=3ORc=4)` instead of a hierarchy of Expr objects for the parenthesed part.
The answer was in the `ParenExpr` rule that should write:

    ParenExpr: '(' expression=Expr ')';

I was fooled by the following writing:

    AbstractRule: ARule | AnotherRule;

Which do not require any assignation.

My take on this is that this behavior is totally unexpected. Is there any rationale on this ?
Else, i have an opinion to share:

I shouldn't be able to write `ParenExpr: '(' Expr ')';` if it yields such type-weird results, unless i explicitly ask for it with a `cast_to_str_when_no_parsing_target` argument.

A probably better fix would be to make such rule an abstract rule. The general definition of abstract rule would then be: *a rule having only one possible target* (because there is only fixed strings elsewhere, or because you have a choice).

In a more general way, i do think that making a grammar just to raise an error if it does not pass is a *special* case. The regular one is to use all parsed data.
As a consequence, i would prefer textx to (by default) throw an error to me when i'm writing something like `Rule: RuleA RuleB`, i.e. a rule that parse while forgetting the parsed data.

Obviously, i am far from expert in grammar related things, and i just happen to run into that problem in my particular case: maybe this is a well-though feature, not a design issue. Would love to hear about the rationale.

**Comment by goto40** [comment]
Hi. Thank you for the example.

Without going too much into details I tried to first reproduce the effect and copied your code into a new unittest file. Somehow it seems to work for me (**maybe I did not correctly understood the error you observed**).

**Could you check** if I correctly tried to reproduce. I tried to formulate **your expectations as asserts**. See: https://github.com/goto40/textX/blob/issue155/tests/functional/regressions/test_issue155.py

I would greatly appreciate to hear from you. Maybe you find time to see the file referenced above.
Best wishes.

P.S: I only tested the current master... Which version were you using?
P.P.S. unittest result: see travis

**Comment by igordejanovic** [comment]
Hi @Aluriak. Just to be sure, which version are you using? stable or master?

**Comment by igordejanovic** [comment]
Ah, I see @goto40 already asked :)  There was a change in handling abstract/match rules in the last few months (only on master) so maybe that makes difference?

**Comment by Aluriak** [comment]
Hi, sorry i should have put the assertions myself.

I am using the stable version i guess (installed using the good old `pip install textx`).

So, i took your [test file](https://github.com/goto40/textX/blob/issue155/tests/functional/regressions/test_issue155.py), @goto40, verified quickly that the assertions were what i expected, made a new virtualenv, pip-installed textx (which means version 1.8.0), add `test_textx_issue155()` at the end of the file and ran it. I got:

     Traceback (most recent call last):
       File "p.py", line 49, in <module>
         test_textx_issue155()
       File "p.py", line 25, in test_textx_issue155
         assert textx_isinstance(m1.expression, METAMODEL['BaseExpr'])
     AssertionError

By adding just before the assertion the line `print(f'"{m1.expression}" of type {type(m1.expression)}')`, i got this before the stacktrace:

    "(" of type <class 'str'>

However, please note that i'm on another PC/configuration than in my first post, so i re-tried my initial code sample, and got:

    (
    <textx:BaseExpr instance at 0x7f1f8deeb198>

Sounds like i got the same problem here as well.

> There was a change in handling abstract/match rules in the last few months (only on master) so maybe that makes difference?

Would totally be our guy !

Don't want to press on you, but that imply that pypi distribution wasn't updated in month ?

**Comment by igordejanovic** [comment]
> Don't want to press on you, but that imply that pypi distribution wasn't updated in month ?

Yup it wasn't. Since we have backward incompatible changes we should be going to version 2.x. I was hoping we could finish TEP-001 for 2.x release but it seems it will take some time. I guess we should go on with some form of semi-scheduled releases and release what is ready.

**Comment by Aluriak** [comment]
In that case, my advice is to follow the *release early, release often* scheme. If you have backward incompatible changes, make them available ASAP so (new) users have less probability to encounter the two incompatible versions. It's not a big deal if 2.0 is not feature-full, but just an API change. You just need to get your user used to the new changes, and able to update their code as soon as they can. Features can arrive later.

Also, as possible, introduce the bug fixes into previous branch.

If you really want to keep the 2.0 for later, you should then get its new (regression) tests into the 1.8.x branch, when they apply (for instance, the *change in handling abstract/match rules in the last few months*).

Do not forget that most users use *only* the pypi distribution (and will not be happy to be forced to bypass that system to just "fix" a bug).

**Comment by goto40** [comment]
@Aluriak, do you have python 3.5? See #159.

**Comment by igordejanovic** [comment]
@Aluriak All valid points. We have to come up to a better release process.

**Comment by Aluriak** [comment]
@goto40 depends of the machine i use. My first post was under 3.6, my second under 3.7.

@igordejanovic I am not an expert of release cycle, i just happen to run into a failing use-case. Focus on what you need to do.

**Comment by goto40** [comment]
@igordejanovic, question: 
```
A: B|C;
B: "B";
C: "C";
```
 * means to me: A is a base of B and C.

What does then the following mean?
```
A: B|C;
B: "B" A;
C: "C";
```
 * A is a base of B,C.
 * B is a A...?

**Comment by goto40** [comment]
(moving comments to PR)

**Comment by igordejanovic** [comment]
@goto40 Good question. In this case all rules are match rules so OO reasoning won't work. We could try to do something like:
```
A: B|C;
B: "B" b=INT;
C: "C" c=INT;
```
And now we have a proper abstract rule `A` and a common rules `B` and `C` and we could reason as `A` is a base of `B` and `C`.

Furthermore we could try to do something similar with another example:
```
A: B|C;
B: "B" A b=INT;
C: "C" c=INT;
```
And here we can say the same for the first rule. But we can't reason anything with the second rule other that it start with `B` and than `A` rule is match. I think `textx visualize` works correct in these cases.
